exercises/dominoes/.meta/src/reference/java/Dominoes.java

```diff
@@ -0,0 +1,75 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+
+class Dominoes {
+	
+	public static void main(String[] args) {
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(3, 1), new Domino(2, 3)};
+		ArrayList<Domino> dominoes = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		ArrayList<Domino> chain = formPartialChain(new ArrayList<Domino>(), dominoes);
+		
+		for(int i = 0; i < chain.size(); i++) {
+			System.out.print("(" + chain.get(i).a + ", " + chain.get(i).b + "), ");
+		}
+	}
+	
+	static ArrayList<Domino> formChain(ArrayList<Domino> inputDominoes) {
+		if(inputDominoes.size() == 0) {
+			return null;
```

Comment:
Please throw an exception on error cases. That's generally considered best practice in Java since it forces the user to consider the error case 🙂

Reply:
Is it really an error though? It is just that there is no valid chain. Also the canonical data file doesn't seem to mention any exceptions. What do you think?

Reply:
Hi @jssander - the canonical data refrains from mentioning exceptions etc. to allow the track maintainers to suggest the best course of action in handling these cases :slightly_smiling_face:

Reply:
Yes, and in Java the best course of action is usually to throw an exception. If you return null then the user of your method has to read your code (or your documentation if you have any) to find out if the code can return null, which is easy to forget. And people shouldn't need to look at the implementation of a method to be able to use it. Also, checking for null all the time can make code look messy. So I'm generally in favour of throwing an exception, even if it's not really an error but just a way of alerting the user to the fact that you haven't been able to return the normal output of the method 🙂 

Another solution could be to return an [`Optional`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html). I think they're also a bit messy and not as nice as exceptions in Java, but they're definitely better than returning null. So you could use an `Optional` if you prefer 🙂

Reply:
Ok. What kind of exception should it be?

Reply:
It's up to you 🙂 You could use an exception which already exists, e.g. `IllegalArgumentException` seems like it would be quite suitable in this case. Or you could make a custom exception like in the triangle exercise 🙂

Reply:
I completely agree with @FridaTveit - you could also check out the [reference solution for error-handling](https://github.com/exercism/java/blob/master/exercises/error-handling/.meta/src/reference/java/ErrorHandling.java) which gives a handful of possibilities :slightly_smiling_face: